### PR TITLE
UI driver

### PR DESCRIPTION
**Pull request by TashreeqPeck**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Byte-compiled / optimized / DLL files
+__pycache__/
+
+# OCR output
+ocr_results*/
```
Comment (TashreeqPeck) [inline .gitignore]: add newline to end of file

File: monkeys/base_monkey.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    def hotkey(self) -> str:
+        """Hotkey to place tower"""
+
+    def can_upgrade(self, path: UpgradePath, money: int | None = None) -> bool:
+        """Check if the monkey can be upgraded"""
+        path_1, path_2 = [_path for _path in UpgradePath if _path is not path]
+        path_0 = self._upgrades[path.value]
+        path_1 = self._upgrades[path_1.value]
+        path_2 = self._upgrades[path_2.value]
+        cost = self.next_upgrade_cost(path)
```
Comment (TashreeqPeck) [inline monkeys/base_monkey.py]: remove cost checking from monkey. It should not have concept of cash

File: monkeys/base_monkey.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        # Handle max upgrades
+        return not (
+            # Can't upgrade more than 5 times
+            path_0 == 4
+            # Can't upgrade more than 2 paths
+            or (path_1 >= 0 and path_2 >= 0)  # type: ignore
+            # Can't upgrade 1 path more than 2 times
+            or (path_0 == 1 and (path_1 > 1 or path_2 > 1))  # type: ignore
+            # Can't upgrade if money is not enough
+            or (cost is not None and money is not None and money < cost)
```
Comment (TashreeqPeck) [inline monkeys/base_monkey.py]: ditto

File: monkeys/base_monkey.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            cost = self._upgrade_cost[path.value][self._upgrades[path.value] + 1]
+        return cost
+
+    def increment_upgrade(self, path: UpgradePath) -> None:
+        """Upgrade the tower"""
+        if not self.can_upgrade(path):
+            raise UpgradeException("Upgrade unsuccessful")
+        self._upgrades[path.value] += 1
+
+    def can_afford(self, money: int) -> bool:
```
Comment (TashreeqPeck) [inline monkeys/base_monkey.py]: ditto

File: monkeys/heroes/__init__.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    def __init__(self, difficulty: Difficulty) -> None:
+        if not hasattr(self, "_base_costs"):
+            raise NotImplementedError(
+                "_base_costs must be implemented in derived class"
+            )
+        self.position: tuple[int, int] | None = None
+        self.cost = self._base_costs[difficulty.value]  # pylint: disable=E1101
+        self.hotkey = "u"
+
+    def can_afford(self, money: int) -> bool:
```
Comment (TashreeqPeck) [inline monkeys/heroes/__init__.py]: remove from class. It should not have a concept of external cash

File: monkeys/base_monkey.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+class UpgradeException(Exception):
+    """Exception for upgrading tower"""
+
+
+class BaseMonkey(ABC):
+    """Base Monkey"""
+
+    def __init__(self, difficulty: Difficulty) -> None:
+        if not hasattr(self, "_upgrade_costs") or not hasattr(self, "_base_costs"):
```
Comment (TashreeqPeck) [inline monkeys/base_monkey.py:27]: refactor into abstract properties

File: monkeys/heroes/__init__.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+logger = logging.getLogger(__name__)
+logger.addHandler(logging.NullHandler())
+# -------------------------------------------------------------------------------------------------
+
+
+class Hero:
+    """Hero Monkey"""
+
+    def __init__(self, difficulty: Difficulty) -> None:
+        if not hasattr(self, "_base_costs"):
```
Comment (TashreeqPeck) [inline monkeys/heroes/__init__.py:21]: refactor into abstract properties

File: ui_controller.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
+"""BloonsTD6 UI control"""
```
Comment (TashreeqPeck) [inline ui_controller.py]: rename to driver

**Review by ghost — APPROVED** [review]

**Pull request merged**